rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,487 @@
+    // TODO: System-wide default token path for Windows
+    PathBuf::from("token.txt")
+}
+
+/// Only called from the GUI Client's build of the IPC service
+///
+/// On Windows, this is wrapped specially so that Windows' service controller
+/// can launch it.
+pub(crate) fn run_only_ipc_service() -> Result<()> {
+    let cli = CliIpcService::parse();
+    match cli.command {
+        CmdIpc::DebugIpcService => run_debug_ipc_service(cli),
+        CmdIpc::IpcService => windows_service::service_dispatcher::start(SERVICE_NAME, ffi_service_run).context("windows_service::service_dispatcher failed. This isn't running in an interactive terminal, right?"),
+    }
+}
+
+fn run_debug_ipc_service(cli: CliIpcService) -> Result<()> {
+    crate::debug_command_setup()?;
+    let rt = tokio::runtime::Runtime::new()?;
+    let mut ipc_service = pin!(ipc_listen(cli));
+    let mut signals = Signals::new()?;
+    rt.block_on(async {
+        std::future::poll_fn(|cx| {
+            match signals.poll(cx) {
+                Poll::Ready(SignalKind::Hangup) => {
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "Impossible, we don't catch Hangup on Windows"
+                    )));
+                }
+                Poll::Ready(SignalKind::Interrupt) => {
+                    tracing::info!("Caught Interrupt signal");
+                    return Poll::Ready(Ok(()));
+                }
+                Poll::Pending => {}
+            }
+
+            match ipc_service.as_mut().poll(cx) {
+                Poll::Ready(Ok(())) => {
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "Impossible, ipc_listencan't return Ok"
+                    )));
+                }
+                Poll::Ready(Err(error)) => {
+                    return Poll::Ready(Err(error).context("ipc_listen failed"));
+                }
+                Poll::Pending => {}
+            }
+
+            Poll::Pending
+        })
+        .await
+    })
+}
+
+// Generates `ffi_service_run` from `service_run`
+windows_service::define_windows_service!(ffi_service_run, windows_service_run);
+
+fn windows_service_run(_arguments: Vec<OsString>) {
+    if let Err(_e) = fallible_windows_service_run() {
+        todo!();
+    }
+}
+
+// Most of the Windows-specific service stuff should go here
+fn fallible_windows_service_run() -> Result<()> {
+    let log_path =
+        crate::known_dirs::ipc_service_logs().context("Can't compute IPC service logs dir")?;
+    std::fs::create_dir_all(&log_path)?;
+    let (layer, _handle) = file_logger::layer(&log_path);
+    let filter = EnvFilter::from_str(SERVICE_RUST_LOG)?;
+    let subscriber = Registry::default().with(layer.with_filter(filter));
+    set_global_default(subscriber)?;
+    tracing::info!(git_version = crate::GIT_VERSION);
+
+    let rt = tokio::runtime::Runtime::new()?;
+    let (shutdown_tx, mut shutdown_rx) = mpsc::channel(1);
```

Comment:
Instead of this channel, you could:

1. Spawn `ipc_listen` onto the runtime
2. Obtain an `AbortHandle` from the returned `JoinHandle`: https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.abort_handle
3. Use the `AbortHandle` inside the event-handler to abort on `Stop`
4. Block the runtime on `JoinHandle::join`.